airbyte_cdk/sources/utils/transform.py

```diff
@@ -225,6 +226,41 @@ def transform(
             logger.warning(self.get_error_message(e))
 
     def get_error_message(self, e: ValidationError) -> str:
-        instance_json_type = python_to_json[type(e.instance)]
-        key_path = "." + ".".join(map(str, e.path))
-        return f"Failed to transform value {repr(e.instance)} of type '{instance_json_type}' to '{e.validator_value}', key path: '{key_path}'"
+        """
+        Construct a sanitized error message from a ValidationError instance.
+        """
+        field_path = ".".join(map(str, e.path))
+        type_structure = self._get_type_structure(e.instance)
+
+        return f"Failed to transform value from type '{type_structure}' to type '{e.validator_value}' at path: '{field_path}'"
+
+    def _get_type_structure(self, input_data: Any, current_depth: int = 0) -> Any:
+        """
+        Get the structure of a given input data for use in error message construction.
+        """
+        # Handle null values
+        if input_data is None:
+            return "null"
+
+        # Avoid recursing too deep
+        if current_depth >= MAX_NESTING_DEPTH:
+            return (
+                "object"
+                if isinstance(input_data, dict)
+                else "array"
+                if isinstance(input_data, list)
+                else python_to_json[type(input_data)]
+            )
+
+        if isinstance(input_data, dict):
+            structure = {
+                key: self._get_type_structure(field_value, current_depth + 1)
+                for key, field_value in input_data.items()
+            }
+
+            if current_depth == 0:
+                return f"object with structure {structure}"
+            return structure
```

Comment:
❓ is the "object with structure" message nessesary?  without it the code is cleaner

```suggestion
        if isinstance(input_data, dict):
            return {
                key: self._get_type_structure(field_value, current_depth + 1)
                for key, field_value in input_data.items()
            }
```